synapse/lib/multislabseqn.py

```diff
@@ -0,0 +1,420 @@
+        if offs >= self.indx - 1:
+            return
+
+        for ridx in range(len(self._ranges) - 1):
+            startidx = self._ranges[ridx]
+
+            if self._openslabs.get(startidx):
+                raise s_exc.SlabInUse(mesg='Attempt to cull while another task is still using it')
+
+            fn = self.slabFilename(self.dirn, startidx)
+            if offs < self._ranges[ridx + 1] - 1:
+                break
+
+            optspath = s_common.switchext(fn, ext='.opts.yaml')
+            try:
+                os.unlink(optspath)
+            except FileNotFoundError:  # pragma: no cover
+                pass
+
+            shutil.rmtree(fn)
+
+            await asyncio.sleep(0)
+
+        self.firstindx = offs + 1
+        self._setFirstIndx(self.tailslab, offs + 1)
+
+        del self._ranges[:ridx]
+
+    def _isTimeToRotate(self, indx):
+        return indx > self._ranges[-1] and indx % self.IDX_PER_SLAB == 0
+
+    async def _makeSlab(self, startidx: int) -> Tuple[s_lmdbslab.Slab, s_slabseqn.SlabSeqn]:
+
+        async with self._openlock:  # Avoid race in two tasks making the same slab
+
+            item = self._openslabs.get(startidx)
+            if item is not None:
+                item[0].incref()
+                return item
+
+            fn = self.slabFilename(self.dirn, startidx)
+
+            slab = await s_lmdbslab.Slab.anit(fn, **self.slabopts)
+            seqn = slab.getSeqn('nexuslog')
+
+            self._openslabs[startidx] = slab, seqn
+
+            def fini():
+                self._openslabs.pop(startidx, None)
+
+            slab.onfini(fini)
+
+            return slab, seqn
+
+    @contextlib.asynccontextmanager
+    async def _getSeqn(self, ridx: int) -> AsyncIterator[s_slabseqn.SlabSeqn]:
+        '''
+        Get the sequence corresponding to an index into self._ranges
+        '''
+        if ridx == len(self._ranges) - 1:
+            assert self.tailslab and self.tailseqn
+            slab, seqn = self.tailslab, self.tailseqn
+
+        elif ridx == self._cacheridx:
+            assert self._cacheslab and self._cacheseqn
+            slab, seqn = self._cacheslab, self._cacheseqn
+
+        else:
+            startidx = self._ranges[ridx]
+
+            self._cacheridx = None
+
+            if self._cacheslab is not None:
+                await self._cacheslab.fini()
+
+            slab, seqn = self._cacheslab, self._cacheseqn = await self._makeSlab(startidx)
+            self._cacheridx = ridx
+
+        slab.incref()
+        try:
+            yield seqn
+        finally:
+            await slab.fini()
+
+    async def add(self, item: Any, indx=None) -> int:
+        '''
+        Add a single item to the sequence.
+        '''
+        advances = True
+
+        if indx is not None:
+            if indx < self.firstindx:
+                raise s_exc.BadLiftValu(mesg=f'indx lower than first index in sequence {self.firstindx}')
```

Comment:
I dont think this is the right exception. BadLiftValu is for storm lifter normalization.

Reply:
ok